libs/providers/multi-provider/src/lib/multi-provider.ts

```diff
@@ -0,0 +1,325 @@
+      }
+
+      providersByName[candidateName] ||= [];
+      providersByName[candidateName].push(constructorProvider.provider);
+    }
+
+    for (const name of Object.keys(providersByName)) {
+      const useIndexedNames = providersByName[name].length > 1;
+      for (let i = 0; i < providersByName[name].length; i++) {
+        const indexedName = useIndexedNames ? `${name}-${i + 1}` : name;
+        this.providerEntriesByName[indexedName] = { provider: providersByName[name][i], name: indexedName };
+        this.providerEntries.push(this.providerEntriesByName[indexedName]);
+        this.statusTracker.wrapEventHandler(this.providerEntriesByName[indexedName]);
+      }
+    }
+
+    // just make sure we don't accidentally modify these later
+    Object.freeze(this.providerEntries);
+    Object.freeze(this.providerEntriesByName);
+  }
+
+  async initialize(context?: EvaluationContext): Promise<void> {
+    await Promise.all(this.providerEntries.map((provider) => provider.provider.initialize?.(context)));
+  }
+
+  async onClose(): Promise<void> {
+    await Promise.all(this.providerEntries.map((provider) => provider.provider.onClose?.()));
+  }
+
+  resolveBooleanEvaluation(
+    flagKey: string,
+    defaultValue: boolean,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<boolean>> {
+    return this.flagResolutionProxy<boolean>(flagKey, 'boolean', defaultValue, context);
+  }
+
+  resolveStringEvaluation(
+    flagKey: string,
+    defaultValue: string,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<string>> {
+    return this.flagResolutionProxy(flagKey, 'string', defaultValue, context);
+  }
+
+  resolveNumberEvaluation(
+    flagKey: string,
+    defaultValue: number,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<number>> {
+    return this.flagResolutionProxy(flagKey, 'number', defaultValue, context);
+  }
+
+  resolveObjectEvaluation<U extends JsonValue>(
+    flagKey: string,
+    defaultValue: U,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<U>> {
+    return this.flagResolutionProxy(flagKey, 'object', defaultValue, context);
+  }
+
+  private async flagResolutionProxy<T extends boolean | string | number | JsonValue>(
+    flagKey: string,
+    flagType: FlagValueType,
+    defaultValue: T,
+    context: EvaluationContext,
+  ): Promise<ResolutionDetails<T>> {
+    const hookContext = this.hookContexts.get(context);
+    const hookHints = this.hookHints.get(context);
+
+    if (!hookContext || !hookHints) {
+      throw new GeneralError(`Hook context not available for evaluation`);
+    }
+
+    let tasks: Promise<boolean>[] = [];
+    let resolutions: ProviderResolutionResult<T>[] = [];
+
+    for (const providerEntry of this.providerEntries) {
+      const task = this.evaluateProviderEntry(
+        flagKey,
+        flagType,
+        defaultValue,
+        providerEntry,
+        hookContext,
+        hookHints,
+        context,
+        resolutions,
+      );
```

Comment:
I am a bit worried about `resolutions` being passed as parameter, but being mutated.
Maybe we can return the the new resolutions `List` and a `boolean` for `shouldEvaluateNext` as a tuple or so.
Especially as the resolutions are even out-of-order in the case of parallel evaluation.

Reply:
Updated this to use a tuple of `[shouldEvaluateNext, resolutionResult]`